### PR TITLE
Validate and sync when activating push for registered device

### PR DESCRIPTION
**Pull request by tcard**
Implements the changes proposed at https://github.com/ably/docs/pull/786.

### CODE REVIEW

**Comment by tcard** [comment]
[Bitrise says ❌](https://app.bitrise.io/build/0f8bb7d4f0b0cdec#?tab=log).

```
[09:57:48]: ▸ Ably_iOS_Tests.PushActivationStateMachine
[09:57:48]: ▸   Activation_state_machine__State_AfterRegistrationSyncFailed__on_Event_CalledActivate__should_fire_SyncRegistrationFailed_event_and_include_device_auth, expected to equal <nil>, got <nil> (use beNil() to match nils)
[09:57:48]: ▸   /Users/vagrant/git/Spec/PushActivationStateMachine.swift:1083
[09:57:48]: ▸   ```
[09:57:48]: ▸                         let deviceAuthorization = request.allHTTPHeaderFields?["X-Ably-DeviceToken"]
[09:57:48]: ▸                         expect(deviceAuthorization).to(equal(deviceIdentityToken))
[09:57:48]: ▸                     }
[09:57:48]: ▸   ```
[09:57:48]: ▸   Activation_state_machine__State_AfterRegistrationSyncFailed__on_Event_CalledActivate__should_fire_RegistrationSynced_event_and_include_device_auth, expected to not be nil, got <nil>
[09:57:48]: ▸   /Users/vagrant/git/Spec/PushActivationStateMachine.swift:1114
[09:57:48]: ▸   ```
[09:57:48]: ▸ 
[09:57:48]: ▸                         expect(stateMachine.rest.device.identityTokenDetails).toNot(beNil())
[09:57:48]: ▸                         expect(stateMachine.current).to(beAKindOf(ARTPushActivationStateWaitingForNewPushDeviceDetails.self))
[09:57:48]: ▸   ```
[09:57:48]: ▸   Activation_state_machine__State_AfterRegistrationSyncFailed__on_Event_CalledActivate__should_fire_RegistrationSynced_event_and_include_device_auth, expected to be nil, got <HtLA+J4n3hrqXOm74RLhh5cC1Fwjv3JSV/urA9Fq/uw=>
[09:57:48]: ▸   /Users/vagrant/git/Spec/PushActivationStateMachine.swift:1146
[09:57:48]: ▸   ```
[09:57:48]: ▸                         let deviceAuthorization = request.allHTTPHeaderFields?["X-Ably-DeviceToken"]
[09:57:48]: ▸                         expect(request.allHTTPHeaderFields?["X-Ably-DeviceSecret"]).to(beNil())
[09:57:48]: ▸                         expect(deviceAuthorization).to(equal(deviceIdentityToken))
[09:57:48]: ▸   ```
[09:57:48]: ▸   Activation_state_machine__State_AfterRegistrationSyncFailed__on_Event_CalledActivate__should_fire_RegistrationSynced_event_and_include_device_auth, expected to equal <nil>, got <nil> (use beNil() to match nils)
[09:57:48]: ▸   /Users/vagrant/git/Spec/PushActivationStateMachine.swift:1147
[09:57:48]: ▸   ```
[09:57:48]: ▸                         expect(request.allHTTPHeaderFields?["X-Ably-DeviceSecret"]).to(beNil())
[09:57:48]: ▸                         expect(deviceAuthorization).to(equal(deviceIdentityToken))
[09:57:48]: ▸                     }
[09:57:48]: ▸   ```
[09:57:48]: ▸   Activation_state_machine__State_AfterRegistrationSyncFailed__on_Event_GotPushDeviceDetails__should_fire_SyncRegistrationFailed_event_and_include_device_auth, expected to equal <nil>, got <nil> (use beNil() to match nils)
[09:57:48]: ▸   /Users/vagrant/git/Spec/PushActivationStateMachine.swift:1288
[09:57:48]: ▸   ```
[09:57:48]: ▸                         let deviceAuthorization = request.allHTTPHeaderFields?["X-Ably-DeviceToken"]
[09:57:48]: ▸                         expect(deviceAuthorization).to(equal(deviceIdentityToken))
[09:57:48]: ▸                     }
[09:57:48]: ▸   ```
[09:57:48]: ▸   Activation_state_machine__State_AfterRegistrationSyncFailed__on_Event_GotPushDeviceDetails__should_fire_RegistrationSynced_event_and_include_device_auth, Unexpected 'identityTokenDetails' is nil
[09:57:48]: ▸   /Users/vagrant/git/Spec/PushActivationStateMachine.swift:1298
[09:57:48]: ▸   ```
[09:57:48]: ▸                         guard let deviceIdentityToken = stateMachine.rest.device.identityTokenDetails?.token else {
[09:57:48]: ▸                             fail("Unexpected 'identityTokenDetails' is nil")
[09:57:48]: ▸                             return
[09:57:48]: ▸   ```
```

**Comment by tcard** [comment]
[Bitrise: ✅](https://app.bitrise.io/build/81ab8abff5f7dafc#?tab=log)

@ricardopereira PTAL

**Comment by ricardopereira** [comment]
I started reviewing this but it has many changes, I need to start over with a fresh head.

BTW, I don't see changes related to `RSH8f`. The `clientId` isn't updated.

**Comment by ricardopereira** [comment]
I reviewed the implementation and it looks good but I think some tests are missing, most of the `RSH8`. Are they supposed to be implemented in another PR?

File: Source/ARTPushActivationState.h [hunk 1]
```diff
@@ -49,10 +49,19 @@ NS_ASSUME_NONNULL_BEGIN
 @interface ARTPushActivationStateWaitingForNewPushDeviceDetails : ARTPushActivationPersistentState
 @end
 
-@interface ARTPushActivationStateWaitingForRegistrationUpdate : ARTPushActivationState
+@interface ARTPushActivationStateWaitingForRegistrationSync : ARTPushActivationState
+
+- (instancetype)initWithMachine:(ARTPushActivationStateMachine *)machine NS_UNAVAILABLE;
++ (instancetype)newWithMachine:(ARTPushActivationStateMachine *)machine NS_UNAVAILABLE;
```
[the remark this answers is not in the document]
Comment (tcard) [reply Source/ARTPushActivationState.h]: I agree, but it was this way already for other states and I didn't want to bloat this PR with orthogonal changes.

File: Source/ARTPushActivationStateMachine.m [hunk 2]
```diff
@@ -256,14 +256,76 @@ - (void)deviceUpdateRegistration:(ARTErrorInfo *)error {
...
+                else {
+                    ARTErrorInfo *missingIdentityTokenError = [ARTErrorInfo createWithCode:0 message:@"Device Identity Token Details is expected"];
+                        [self sendEvent:[ARTPushActivationEventSyncRegistrationFailed newWithError:missingIdentityTokenError]];
+                }
+            }];
+        });
+        return;
+    }
+
+    void (^doDeviceSync)(void) = ^{
```
[the remark this answers is not in the document]
Comment (tcard) [reply Source/ARTPushActivationStateMachine.m]: We only use weak references to break long-lived reference cycles only. There's no reference cycle here, as, while the block references `self`, `self` doesn't reference the block.

Also, using a weak ref introduces complexity, as then the block, when called, could find its ref to `self` in two "modes": nil and non-nil, while by using a strong ref we only need to care about the non-nil case.

We didn't have a strongly defined policy for weak vs. strong refs until [this issue](https://github.com/ably/ably-cocoa/issues/879) came along. See that issue, this [associated PR](https://github.com/ably/ably-cocoa/pull/883) (which was abandoned, but partially lived on as a318d720699bfc221537797e9c2e6117f6f2025f), and maybe a [more general intro+definition of our memory management policy](https://gitlab.com/snippets/1882500#switch-to-strong-references-where-possible).

PS. I realize this isn't how things should be documented. And I'm definitely not comfortable about how much of this knowledge is only in @ricardopereira's and my heads. Perhaps I could allocate some time to write a "development guide" to clearly state this and other practices we follow, the overall design, etc.?

**Comment by tcard** [comment]
@QuintinWillison 

> * How are you testing this?

See the changes to Spec/PushActivationStateMachine.swift. The diff is messy because I had to refactor it a bit, but basically, for every item in the spec, we have an associated test. For the state machine, that means a test for each state+event+side effect.

> 2\. Could you consider using more `const` in new code if possible, please - I'm a fan of making method arguments and local variables, in particular when they are pointers, constant so that the compiler can spot if they get trampled and the intent of their use is clearer to future maintainers.

Sure. I know we talked about it but I'm just not used to it. I'll sprinkle some of these over.

@ricardopereira 

> I reviewed the implementation and it looks good but I think some tests are missing, most of the `RSH8`. Are they supposed to be implemented in another PR?

True, I focused on the state machine changes, but RSH8 should be tested as well (I believe all items are already implemented).

File: Source/ARTPushActivationState.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (QuintinWillison) [reply Source/ARTPushActivationState.h]: Fair point. As you were. 😄

File: Source/ARTPushActivationStateMachine.m (hunk 2 above)
[the remark this answers is not in the document]
Comment (QuintinWillison) [reply Source/ARTPushActivationStateMachine.m]: It certainly looks like you've put a lot of thought in to this so happy to leave as it is.

File: Source/ARTPushActivationStateMachine.m [hunk 3]
```diff
@@ -256,14 +256,76 @@ - (void)deviceUpdateRegistration:(ARTErrorInfo *)error {
...
+        // Asynchronous HTTP request
+        NSString *path = [@"/push/deviceRegistrations" stringByAppendingPathComponent:local.id];
+        NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:[NSURL URLWithString:path]];
+        request.HTTPMethod = @"PUT";
+        request.HTTPBody = [[self->_rest defaultEncoder] encodeDeviceDetails:local error:nil];
+        [request setValue:[[self->_rest defaultEncoder] mimeType] forHTTPHeaderField:@"Content-Type"];
+
+        [[self->_rest logger] debug:__FILE__ line:__LINE__ message:@"%@: sync device with request %@", NSStringFromClass(self.class), request];
+        [self->_rest executeRequest:request withAuthOption:ARTAuthenticationOn completion:^(NSHTTPURLResponse *response, NSData *data, NSError *error) {
```
[the remark this answers is not in the document]
Comment (QuintinWillison) [reply Source/ARTPushActivationStateMachine.m]: Happy to leave as is, given your [earlier explanation](https://github.com/ably/ably-cocoa/pull/974#discussion_r373986551).

**Comment by tcard** [comment]
[Bitrise: ✅](https://app.bitrise.io/build/f3bd316439820a1e#?tab=log)

@ricardopereira @QuintinWillison PTAL

**Pull request merged**